ocaml/xapi/xapi_pusb.ml

```diff
@@ -31,6 +32,72 @@ let create ~__context ~uSB_group ~host ~other_config ~path
   debug "PUSB ref='%s' created" (Ref.string_of pusb);
   pusb
 
+let scan_start ~__context =
+  let host = Helpers.get_localhost ~__context in
+  let known_usb =
+    Db.PUSB.get_all_records ~__context
+    |> List.filter (fun (rf, rc) -> rc.API.pUSB_host = host)
+    |> get_known_usb
+  in
+  let local_usb =
+    get_usbs ~__context
+    |> get_local_usb
+  in
+  (* Create the newly added pusbs *)
+  USBSet.iter (fun s -> let self = create ~__context ~uSB_group:(Ref.null) ~host ~other_config:[] ~path:s.USB.path ~vendor_id:s.USB.vendor_id
+                   ~vendor_desc:s.USB.vendor_desc ~product_id:s.USB.product_id ~product_desc:s.USB.product_desc ~serial:s.USB.serial
+                   ~version:s.USB.version ~description:s.USB.description in
+               let group = Xapi_pusb_helpers.find_or_create ~__context self in
+               Db.PUSB.set_USB_group ~__context ~self ~value:group
+  ) (USBSet.diff local_usb known_usb);
+
+  let get_obsolete_pusb s = List.filter (fun (rf, rc) -> rc.API.pUSB_path = s.USB.path && rc.API.pUSB_vendor_id = s.USB.vendor_id
+                            && rc.API.pUSB_product_id = s.USB.product_id && rc.API.pUSB_serial = s.USB.serial) (Db.PUSB.get_all_records ~__context) in
+  (* Delete the obsolete pusbs *)
+  USBSet.iter (fun s ->
```

Comment:
I don't understand this code. Why does this need get_obsolete_pusb? Isn't the difference between known_usb and local_usb containing all usb devices that need to be removed?

Reply:
Yes, (USBSet.diff known_usb local_usb) contains all usb devices to remove. Here get_obsolete_pusb is to transform USB.t to PUSB records to destroy.

Reply:
This looks inefficient. For every element in the set the code calls Db.PUSB.get_all_records to filter out one element, which then is removed.

Why not:

```
List.filter (fun (rf, rc) -> (test element is in set) (Db.PUSB.get_all_records) 
|> destroy 
```

Reply:
hmm, yes. It is indeed inefficient.

Reply:
Have refined. Please take a look if it is ok. Thanks